hashin.py

```diff
@@ -228,6 +219,69 @@ def run_packages(
         if verbose:
             _verbose("Editing", file)
 
+    return 0
+
+
+class InteractiveAll(Exception):
+    """When the user wants to say yes to ALL package updates."""
+
+
+class InteractiveQuit(Exception):
+    """When the user wants to stop the interactive update questions entirely."""
+
+
+def interactive_upgrade_request(package, old_version, new_version, print_header=False):
+    def print_version(v):
+        return str(v).replace("==", "").ljust(15)
+
+    if print_header:
+        print(
+            "PACKAGE".ljust(30),
+            print_version("YOUR VERSION"),
+            print_version("NEW VERSION"),
+        )
+
+    def print_line(checkbox=None):
+        if checkbox is None:
+            checkboxed = "?"
+        elif checkbox:
+            checkboxed = "✓"
+        else:
+            checkboxed = "✘"
+        print(
+            package.ljust(30),
+            print_version(old_version),
+            print_version(new_version),
+            checkboxed,
+        )
+
+    print_line()
+
+    def clear_line():
+        sys.stdout.write("\033[F")  # Cursor up one line
+        sys.stdout.write("\033[K")  # Clear to the end of line
+
+    def ask():
+        answer = input("Update? [Y/n/a/q]: ").lower().strip()
+        if answer == "n":
+            clear_line()
+            clear_line()
+            print_line(False)
+            return False
+        if answer == "a":
+            clear_line()
+            raise InteractiveAll
+        if answer == "q":
+            raise InteractiveQuit
+        if answer == "y" or answer == "" or answer == "yes":
+            clear_line()
+            clear_line()
+            print_line(True)
+            return True
+        return ask()
```

Comment:
It might be better to put this in a loop instead of recursing. I think that makes the repetition easier to read.

Reply:
Point taken. I can see how that could be used to achieve the same thing. But I kinda like it as a function like that. Perhaps it fits my brain better.